dynamic shapes mismatch errors

avikchaudhuri opened the pull request:
Summary: When PyTree detects a structural mismatch between inputs and dynamic shapes, the error messages are quite horrible. This PR fixes these error messages by adding, for each kind of error, the path to the point where the error happens and an actionable reason for the error.

Test Plan: added test with several cases

Differential Revision: D60956976

Skylion007 commented on torch/export/dynamic_shapes.py:
```diff
@@ -612,22 +617,85 @@ def is_leaf(t):
+                if type_ in (list, tuple):
+                    assert context is None
+                    return SequenceKey(i)
+                raise AssertionError(f"Did not expect type {type_}")
```
FYI, should look into 'f{value=}' f-string formatter. Can give informative error messages with a nice shorthand.

avikchaudhuri committed:
dynamic shapes mismatch errors (#132982)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/132982

When PyTree detects a structural mismatch between inputs and dynamic shapes, the error messages are quite horrible. This PR fixes these error messages by adding, for each kind of error, the path to the point where the error happens and an actionable reason for the error.

Test Plan: added test with several cases

Reviewed By: yushangdi

Differential Revision: D60956976
